### ISSUE TITLE
[bug-WinOS] Segmentation fault in `example/resnet50_onnx`

### ISSUE THREAD

**Issue by kelreeeeey** [ariaghora/anvil#15]
I found the bug in 2 places 
- [`tensor.conv2d_xwb`](https://github.com/ariaghora/anvil/blob/e5e117a69a1dd958866cd87e6c532a38208ff5fb/anvil/tensor/conv.odin#L1416-L1434), and
- [`tensor.reshape_bhwc_to_bchw`](https://github.com/ariaghora/anvil/blob/e5e117a69a1dd958866cd87e6c532a38208ff5fb/anvil/tensor/conv.odin#L1226-L1249)

The bug is similar to [`tensor.softmax_last_dim_inplace`](https://github.com/ariaghora/anvil/issues/2), alignment issue when using `#simd`

I've tried to change the implementation in similar manner as previous issue, and the bug goes away

<details>

```odin
reshape_bhwc_to_bchw :: proc(
	tensor: ^Tensor($T),
	batch, height, width, channels: uint,
	allocator := context.allocator,
) -> ^Tensor(T) {
	result := tensor_alloc(T, []uint{batch, channels, height, width}, true, allocator)

	src := tensor.data
	dst := result.data
	hw := height * width

	#no_bounds_check {
		when T == f32 {
			for b in 0 ..< batch {
				src_batch := src[b * hw * channels:]
				dst_batch := dst[b * channels * hw:]

				// Process 4 spatial positions at once
				hw_idx := uint(0)
				for ; hw_idx + 4 <= hw; hw_idx += 4 {
					src_base := hw_idx * channels

					// Process 4 channels at a time
					c := uint(0)
					for ; c + 4 <= channels; c += 4 {
						// Load 4x4 block, 4 spatial × 4 channels
						// These are already consecutive, so use SIMD loads
						row0 := (^#simd[4]f32)(&src_batch[src_base +                c])^
						row1 := (^#simd[4]f32)(&src_batch[src_base +     channels + c])^
						row2 := (^#simd[4]f32)(&src_batch[src_base + 2 * channels + c])^
						row3 := (^#simd[4]f32)(&src_batch[src_base + 3 * channels + c])^

						// Transpose 4x4 matrix using shuffles
						// Step 1: Interleave pairs
						tmp0 := simd.shuffle(row0, row1, 0, 1, 4, 5) // [r0[0], r0[1], r1[0], r1[1]]
						tmp1 := simd.shuffle(row0, row1, 2, 3, 6, 7) // [r0[2], r0[3], r1[2], r1[3]]
						tmp2 := simd.shuffle(row2, row3, 0, 1, 4, 5) // [r2[0], r2[1], r3[0], r3[1]]
						tmp3 := simd.shuffle(row2, row3, 2, 3, 6, 7) // [r2[2], r2[3], r3[2], r3[3]]

						// Step 2: Final shuffle to get transposed vectors
						dst_c0 := simd.shuffle(tmp0, tmp2, 0, 2, 4, 6) // All channel 0s
						dst_c1 := simd.shuffle(tmp0, tmp2, 1, 3, 5, 7) // All channel 1s
						dst_c2 := simd.shuffle(tmp1, tmp3, 0, 2, 4, 6) // All channel 2s
						dst_c3 := simd.shuffle(tmp1, tmp3, 1, 3, 5, 7) // All channel 3s

						#unroll for i in 0 ..< 4 {

							dst_batch[c       * hw + hw_idx + uint(i+0)] = simd.extract(dst_c0, i+0)
							dst_batch[c       * hw + hw_idx + uint(i+1)] = simd.extract(dst_c0, i+1)
							dst_batch[c       * hw + hw_idx + uint(i+2)] = simd.extract(dst_c0, i+2)
							dst_batch[c       * hw + hw_idx + uint(i+3)] = simd.extract(dst_c0, i+3)

							dst_batch[(c + 1) * hw + hw_idx + uint(i+0)] = simd.extract(dst_c1, i+0)
							dst_batch[(c + 1) * hw + hw_idx + uint(i+1)] = simd.extract(dst_c1, i+1)
							dst_batch[(c + 1) * hw + hw_idx + uint(i+2)] = simd.extract(dst_c1, i+2)
							dst_batch[(c + 1) * hw + hw_idx + uint(i+3)] = simd.extract(dst_c1, i+3)

							dst_batch[(c + 2) * hw + hw_idx + uint(i+0)] = simd.extract(dst_c2, i+0)
							dst_batch[(c + 2) * hw + hw_idx + uint(i+1)] = simd.extract(dst_c2, i+1)
							dst_batch[(c + 2) * hw + hw_idx + uint(i+2)] = simd.extract(dst_c2, i+2)
							dst_batch[(c + 2) * hw + hw_idx + uint(i+3)] = simd.extract(dst_c2, i+3)

							dst_batch[(c + 3) * hw + hw_idx + uint(i+0)] = simd.extract(dst_c3, i+0)
							dst_batch[(c + 3) * hw + hw_idx + uint(i+1)] = simd.extract(dst_c3, i+1)
							dst_batch[(c + 3) * hw + hw_idx + uint(i+2)] = simd.extract(dst_c3, i+2)
							dst_batch[(c + 3) * hw + hw_idx + uint(i+3)] = simd.extract(dst_c3, i+3)

						}
						// (^#simd[4]f32)(&dst_batch[c       * hw + hw_idx])^ = dst_c0
						// (^#simd[4]f32)(&dst_batch[(c + 1) * hw + hw_idx])^ = dst_c1
						// (^#simd[4]f32)(&dst_batch[(c + 2) * hw + hw_idx])^ = dst_c2
						// (^#simd[4]f32)(&dst_batch[(c + 3) * hw + hw_idx])^ = dst_c3
					}

					// Handle remainder channels
					for ; c < channels; c += 1 {
						vals := #simd[4]f32 {
							src_batch[src_base + c],
							src_batch[src_base + channels + c],
							src_batch[src_base + 2 * channels + c],
							src_batch[src_base + 3 * channels + c],
						}
						(^#simd[4]f32)(&dst_batch[c * hw + hw_idx])^ = vals
					}
				}

				// Handle remainder spatial positions
				for ; hw_idx < hw; hw_idx += 1 {
					src_offset := hw_idx * channels

					c := uint(0)
					for ; c + 4 <= channels; c += 4 {
						vals := #simd[4]f32 {
							src_batch[src_offset + c],
							src_batch[src_offset + c + 1],
							src_batch[src_offset + c + 2],
							src_batch[src_offset + c + 3],
						}

						// Destinations are NOT consecutive (strided by hw),
						// so we're stuck with scalar stores  ¯\_(ツ)_/¯
						dst_batch[c * hw + hw_idx] = simd.extract(vals, 0)
						dst_batch[(c + 1) * hw + hw_idx] = simd.extract(vals, 1)
						dst_batch[(c + 2) * hw + hw_idx] = simd.extract(vals, 2)
						dst_batch[(c + 3) * hw + hw_idx] = simd.extract(vals, 3)
					}

					for ; c < channels; c += 1 {
						dst_batch[c * hw + hw_idx] = src_batch[src_offset + c]
					}
				}
			}
		} else {
			// Non-SIMD, tiled
			for b in 0 ..< batch {
				for c_tile := uint(0); c_tile < channels; c_tile += TILE_SIZE {
					c_end := min(c_tile + TILE_SIZE, channels)

					for h in 0 ..< height {
						for w in 0 ..< width {
							src_base := b * height * width * channels + (h * width + w) * channels
							dst_base := b * channels * height * width + h * width + w

							if c_end - c_tile == TILE_SIZE {
								#unroll for i in 0 ..< TILE_SIZE {
									dst[dst_base + (c_tile + uint(i)) * height * width] =
										src[src_base + c_tile + uint(i)]
								}
							} else {
								for c := c_tile; c < c_end; c += 1 {
									dst[dst_base + c * height * width] = src[src_base + c]
								}
							}
						}
					}
				}
			}
		}
	}

	return result
}

// ...

conv2d_xwb :: proc(
	input: ^Tensor($T), // (B, C_in, H, W)
	kernel: ^Tensor(T), // (C_out, C_in, K_h, K_w)
	bias: Maybe(^Tensor(T)), // (C_out, C_in, K_h, K_w)
	stride, dilation, padding, groups: uint,
	allocator := context.allocator,
	loc := #caller_location,
) -> (
	out: ^tensor.Tensor(T),
) {
	conv2d_xwb_trace := trace.TRACE_FUNCTION("conv2d_xwb")
	defer trace.end_scoped_trace(conv2d_xwb_trace)

	if groups == 1 {
		conv2d_single_trace := trace.TRACE_SECTION("conv2d_single")
		defer trace.end_scoped_trace(conv2d_single_trace)
		out = tensor.conv2d_single(input, kernel, stride, dilation, padding, allocator, loc)
	} else {
		conv2d_grouped_trace := trace.TRACE_SECTION("conv2d_grouped")
		defer trace.end_scoped_trace(conv2d_grouped_trace)
		out = tensor.conv2d_grouped(
			input,
			kernel,
			groups,
			stride,
			dilation,
			padding,
			allocator,
			loc,
		)
	}

	// Add bias if present
	if bias, has_bias := bias.?; has_bias {
		add_bias_trace := trace.TRACE_SECTION("add_bias")
		defer trace.end_scoped_trace(add_bias_trace)

		batch_size := out.shape[0]
		out_channels := out.shape[1]
		spatial_size := out.shape[2] * out.shape[3]

		#no_bounds_check when T == f32 {
			for b in 0 ..< batch_size {
				for c in 0 ..< out_channels {
					base_idx := (b * out_channels + c) * spatial_size
					bias_val := bias.data[c]

					when ODIN_OS == .Darwin {
						simd_backend.vsaddf_batch(
							out.data[base_idx:base_idx + spatial_size],
							out.data[base_idx:base_idx + spatial_size],
							&bias_val,
						)
					} else {
						bias_vec := #simd[4]f32{bias_val, bias_val, bias_val, bias_val}
						i := uint(0)
						for ; i + 8 <= spatial_size; i += 8 {
							// vals0 := (^#simd[4]f32)(&out.data[base_idx + i])^
							vals0 := #simd[4]f32 {
								out.data[base_idx + i + 0],
								out.data[base_idx + i + 1],
								out.data[base_idx + i + 2],
								out.data[base_idx + i + 3],
							}
							// vals1 := (^#simd[4]f32)(&out.data[base_idx + i + 4])^
							vals1 := #simd[4]f32 {
								out.data[base_idx + i + 4 + 0],
								out.data[base_idx + i + 4 + 1],
								out.data[base_idx + i + 4 + 2],
								out.data[base_idx + i + 4 + 3],
							}

							vals0 += bias_vec
							vals1 += bias_vec

							// (^#simd[4]f32)(&out.data[base_idx + i])^ = vals0
							// (^#simd[4]f32)(&out.data[base_idx + i + 4])^ = vals1

							#unroll for i in 0 ..< 4 {
								out.data[base_idx + uint(i     + 0)] = simd.extract(vals0, i)
								out.data[base_idx + uint(i     + 1)] = simd.extract(vals0, i)
								out.data[base_idx + uint(i     + 2)] = simd.extract(vals0, i)
								out.data[base_idx + uint(i     + 3)] = simd.extract(vals0, i)
								out.data[base_idx + uint(i + 4 + 0)] = simd.extract(vals1, i)
								out.data[base_idx + uint(i + 4 + 1)] = simd.extract(vals1, i)
								out.data[base_idx + uint(i + 4 + 2)] = simd.extract(vals1, i)
								out.data[base_idx + uint(i + 4 + 3)] = simd.extract(vals1, i)
							}
						}

						for ; i + 4 <= spatial_size; i += 4 {
							// vals := (^#simd[4]f32)(&out.data[base_idx + i])^
							vals := #simd[4]f32 {
								out.data[base_idx + uint(i + 0)],
								out.data[base_idx + uint(i + 1)],
								out.data[base_idx + uint(i + 2)],
								out.data[base_idx + uint(i + 3)],
							}
							vals += bias_vec
							// (^#simd[4]f32)(&out.data[base_idx + i])^ = vals
							#unroll for i in 0 ..< 4 {
								out.data[base_idx + uint(i     + 0)] = simd.extract(vals, i)
								out.data[base_idx + uint(i     + 1)] = simd.extract(vals, i)
								out.data[base_idx + uint(i     + 2)] = simd.extract(vals, i)
								out.data[base_idx + uint(i     + 3)] = simd.extract(vals, i)
							}
						}

						for ; i < spatial_size; i += 1 {
							out.data[base_idx + i] += bias_val
						}
					}
				}
			}
		} else {
			// Scalar fallback
			for b in 0 ..< batch_size {
				for c in 0 ..< out_channels {
					bias_val := bias.data[c]
					base_idx := (b * out_channels + c) * spatial_size
					for i in 0 ..< spatial_size {
						out.data[base_idx + i] += bias_val
					}
				}
			}
		}
	}
	return
}

```

</details>

However, with this image

<details>

![Image](https://github.com/user-attachments/assets/40e589a3-d798-4d88-8b0c-95a0680a79a3)

</details>

The program gave me

```bash
$ odin build examples/resnet50_onnx -o:speed -no-bounds-check  && ./resnet50_onnx.exe "../llamas.jpeg"
Producer Name    :  pytorch
Producer Version :  2.7.0
Opset Version    :  11
Inference time   :  192.6096ms
Predicted class  :  geyser
```

**Comment by ariaghora** [comment]
Thanks. But I'd like to actually solve the actual problem rather than patching this way. Besides, I've tried this way during fixing ubuntu version, but it made the code much slower. Let's see what I can do.

**Comment by ariaghora** [comment]
<img width="1678" height="322" alt="Image" src="https://github.com/user-attachments/assets/ad8c1538-dd39-4832-bc3f-b5ba62f0b835" />

Yep, confirmed on ubuntu too. Looking into this more

**Comment by ariaghora** [comment]
All tensor buffers are aligned. But real issue is, the stride pattern.
Even if `dst_batch` starts at a perfectly aligned address, when we jump by `hw` (height × width) between channels, we break alignment if `hw` isn't a multiple of 4. Our options:

- Padding the tensors to maintain alignment (friggin lazy, nope)
- Branching on `hw % 4` once and having two code paths: simd way and your way (extract one by one)

On `reshape_bhwc_to_bchw`:

```odin
	if hw % 4 != 0 {
		// extract one by one
	} else { 
		// current simd way
	}
```
I will consider branching and use your way in case of nonaligned `hw`.

This might be the issue for the `conv2_xwb`.
That's why folks, use pretty numbers to determine sizes (image, kernels, etc) in neural nets `¯\_(ツ)_/¯`

**Comment by ariaghora** [comment]
Fixed on 2b3c816 on main branch. Can you check if this resolves the issue at your end, @kelreeeeey ?
